adapter/kubernetes/kubernetes.go

```diff
@@ -260,17 +289,76 @@ func addPodValues(m map[string]interface{}, prefix string, params config.Params,
 	if p == nil {
 		return
 	}
-	m[valueName(prefix, params.LabelsValueName)] = p.Labels
-	m[valueName(prefix, params.PodNameValueName)] = p.Name
-	m[valueName(prefix, params.NamespaceValueName)] = p.Namespace
-	m[valueName(prefix, params.ServiceAccountValueName)] = p.Spec.ServiceAccountName
-	m[valueName(prefix, params.PodIpValueName)] = p.Status.PodIP
-	m[valueName(prefix, params.HostIpValueName)] = p.Status.HostIP
+	if len(p.Labels) > 0 {
+		m[valueName(prefix, params.LabelsValueName)] = p.Labels
+	}
+	if len(p.Name) > 0 {
+		m[valueName(prefix, params.PodNameValueName)] = p.Name
+	}
+	if len(p.Namespace) > 0 {
+		m[valueName(prefix, params.NamespaceValueName)] = p.Namespace
+	}
+	if len(p.Spec.ServiceAccountName) > 0 {
+		m[valueName(prefix, params.ServiceAccountValueName)] = p.Spec.ServiceAccountName
+	}
+	if len(p.Status.PodIP) > 0 {
+		m[valueName(prefix, params.PodIpValueName)] = p.Status.PodIP
+	}
+	if len(p.Status.HostIP) > 0 {
+		m[valueName(prefix, params.HostIpValueName)] = p.Status.HostIP
+	}
 	if app, found := p.Labels[params.PodLabelForService]; found {
-		m[valueName(prefix, params.ServiceValueName)] = app
+		n, err := canonicalName(app, p.Namespace, params.ClusterDomainName)
+		if err == nil {
+			m[valueName(prefix, params.ServiceValueName)] = n
+		}
 	}
 }
 
 func valueName(prefix, value string) string {
 	return fmt.Sprintf("%s%s", prefix, value)
 }
+
+// name format examples that can be currently canonicalized:
+//
+// "hello:80",
+// "hello",
+// "hello.default:80",
+// "hello.default",
+// "hello.default.svc:80",
+// "hello.default.svc",
+// "hello.default.svc.cluster:80",
+// "hello.default.svc.cluster",
+// "hello.default.svc.cluster.local:80",
+// "hello.default.svc.cluster.local",
+func canonicalName(service, namespace, clusterDomain string) (string, error) {
+	if len(service) == 0 {
+		return "", errors.New("invalid service name: cannot be empty")
+	}
+	// remove any port suffixes (ex: ":80")
+	splits := strings.SplitN(service, ":", 2)
+	s := splits[0]
+	if len(s) == 0 {
+		return "", errors.New("invalid service name: starts with ':'")
+	}
+	// error on ip addresses for now
+	if ip := net.ParseIP(s); ip != nil {
+		return "", errors.New("invalid service name: cannot canonicalize ip addresses at this time")
+	}
+	parts := strings.SplitN(s, ".", 5)
```

Comment:
The kube dns domain specified the suffix = `svc.cluster.local`, but it could be longer than that. mixer get this as a command line parameter.